web/client/src/library/components/sourceList/SourceList.tsx

```diff
@@ -61,51 +67,142 @@ export default function SourceList<
-            }}
-          >
-            No Results Found
-          </li>
-        )}
-        {filtered.map(item => {
-          const id = (item as Record<string, string>)[by]!
-          const name = isNil(byName)
-            ? ''
-            : (item as Record<string, string>)?.[byName] ?? ''
-          const description = isNil(byDescription)
-            ? undefined
-            : (item as Record<string, string>)?.[byDescription] ?? undefined
+  const rowVirtualizer = useVirtualizer({
+    count: filtered.length,
+    getScrollElement: () => scrollableAreaRef.current,
+    estimateSize: () => 28,
+  })
 
-          return (
+  const scrollToItem = ({
+    itemIndex,
+    isSmoothScroll = true,
+  }: {
+    itemIndex: number
+    isSmoothScroll?: boolean
+  }): void => {
+    rowVirtualizer.scrollToIndex(itemIndex, {
+      align: 'center',
+      behavior: isSmoothScroll ? 'smooth' : 'auto',
+    })
+  }
+
+  // the index of the active item in the filtered list
+  const filteredItemIndex = useMemo(() => {
+    const activeItem = items[activeItemIndex]
+    const filteredIndex = filtered.findIndex(
+      filteredItem => filteredItem.name === activeItem?.name,
+    )
+    return filteredIndex
+  }, [filtered, activeItemIndex])
+
+  /**
+   * The return button should appear when the
+   * active item is available in the list (not
+   * filtered out) and it is not in the visible
+   * range of the virtualized list
+   */
+  const shouldShowReturnButton = useMemo(() => {
+    if (filteredItemIndex > -1) {
+      // active item is in the currently filtered list
+      const range = rowVirtualizer.range
+
+      // check if it's outside the visible range
+      if (
+        isNotNil(range) &&
+        (range.startIndex > filteredItemIndex ||
+          range?.endIndex < filteredItemIndex)
+      ) {
+        return true
+      }
+    }
+    return false
+  }, [filteredItemIndex, rowVirtualizer.range])
+
+  // scroll to the active item when the activeItemIndex changes
+  useEffect(() => {
+    if (
+      filteredItemIndex > -1 &&
+      isNotNil(rowVirtualizer.range) &&
+      (rowVirtualizer.range.startIndex > filteredItemIndex ||
+        rowVirtualizer.range.endIndex < filteredItemIndex)
+    ) {
+      scrollToItem({ itemIndex: filteredItemIndex, isSmoothScroll: false })
+    }
+  }, [activeItemIndex])
+
+  return (
+    <div className={clsx('flex flex-col w-full h-full relative', className)}>
+      {shouldShowReturnButton && (
+        <Button
+          className="absolute right-0 top-0 z-10 text-ellipsis !block overflow-hidden no-wrap max-w-[90%]"
+          onClick={() => scrollToItem({ itemIndex: filteredItemIndex })}
+          size="sm"
+          variant="neutral"
+        >
+          Scroll to {items[activeItemIndex]?.name}
+        </Button>
+      )}
+      <div
+        className="p-2 h-full overflow-auto hover:scrollbar scrollbar--horizontal scrollbar--vertical"
+        ref={scrollableAreaRef}
+      >
+        <div
```

Comment:
can this still be `ul` or we need to change `li`s to maybe `div`s or `span`s

Reply:
yes thank you, I meant to revisit that after I got this all working with divs (because I had to add an extra wrapper inside) and I forgot. I'm switching it back to a list!